File: svgx_engine/security/authentication.py
"""
Authentication and Authorization Services for Arxos.

This module implements enterprise-grade authentication and authorization
with support for RBAC (Role-Based Access Control) and ABAC (Attribute-Based Access Control).
"""

import jwt
import hashlib
import secrets
import time
from datetime import datetime, timedelta
from typing import Dict, List, Optional, Any, Union
from dataclasses import dataclass
from enum import Enum
import bcrypt
from cryptography.fernet import Fernet


class UserRole(Enum):
    """User roles for RBAC implementation."""
    ADMIN = "admin"
    ENGINEER = "engineer"
    VIEWER = "viewer"
    CONTRACTOR = "contractor"
    GUEST = "guest"


class Permission(Enum):
    """Permissions for resource access control."""
    READ = "read"
    WRITE = "write"
    DELETE = "delete"
    MANAGE_USERS = "manage_users"
    MANAGE_SYSTEM = "manage_system"
    EXPORT_DATA = "export_data"
    IMPORT_DATA = "import_data"
    VIEW_AUDIT_LOGS = "view_audit_logs"


@dataclass
class User:
    """User entity with attributes for ABAC."""
    id: str
    username: str
    email: str
    roles: List[UserRole]
    attributes: Dict[str, Any]
    is_active: bool = True
    created_at: datetime = None
    last_login: datetime = None


@dataclass
class Resource:
    """Resource entity for access control."""
    id: str
    type: str
    attributes: Dict[str, Any]
    owner_id: str
    created_at: datetime = None


class AuthService:
    """Main authentication service with JWT token management."""
    
    def __init__(self, secret_key: str, algorithm: str = "HS256"):
        self.secret_key = secret_key
        self.algorithm = algorithm
        self.token_expiry = timedelta(hours=24)
        self.refresh_token_expiry = timedelta(days=7)
        
    def hash_password(self, password: str) -> str:
        """Hash password using bcrypt."""
        salt = bcrypt.gensalt()
        return bcrypt.hashpw(password.encode('utf-8'), salt).decode('utf-8')
    
    def verify_password(self, password: str, hashed_password: str) -> bool:
        """Verify password against hash."""
        return bcrypt.checkpw(password.encode('utf-8'), hashed_password.encode('utf-8'))
    
    def generate_token(self, user: User) -> str:
        """Generate JWT token for user."""
        payload = {
            'user_id': user.id,
            'username': user.username,
            'roles': [role.value for role in user.roles],
            'exp': datetime.utcnow() + self.token_expiry,
            'iat': datetime.utcnow(),
            'jti': secrets.token_urlsafe(32)
        }
        return jwt.encode(payload, self.secret_key, algorithm=self.algorithm)
    
    def verify_token(self, token: str) -> Optional[Dict]:
        """Verify and decode JWT token."""
        try:
            payload = jwt.decode(token, self.secret_key, algorithms=[self.algorithm])
            return payload
        except jwt.ExpiredSignatureError:
            raise ValueError("Token has expired")
        except jwt.InvalidTokenError:
            raise ValueError("Invalid token")
    
    def refresh_token(self, refresh_token: str) -> str:
        """Generate new access token from refresh token."""
        try:
            payload = jwt.decode(refresh_token, self.secret_key, algorithms=[self.algorithm])
            # Create new user object from payload
            user = User(
                id=payload['user_id'],
                username=payload['username'],
                email="",  # Not stored in token for security
                roles=[UserRole(role) for role in payload['roles']],
                attributes={}
            )
            return self.generate_token(user)
        except jwt.InvalidTokenError:
            raise ValueError("Invalid refresh token")


class RBACService:
    """Role-Based Access Control implementation."""
    
    def __init__(self):
        self.role_permissions = {
            UserRole.ADMIN: [
                Permission.READ, Permission.WRITE, Permission.DELETE,
                Permission.MANAGE_USERS, Permission.MANAGE_SYSTEM,
                Permission.EXPORT_DATA, Permission.IMPORT_DATA,
                Permission.VIEW_AUDIT_LOGS
            ],
            UserRole.ENGINEER: [
                Permission.READ, Permission.WRITE, Permission.EXPORT_DATA,
                Permission.IMPORT_DATA
            ],
            UserRole.VIEWER: [
                Permission.READ
            ],
            UserRole.CONTRACTOR: [
                Permission.READ, Permission.LIMITED_WRITE
            ],
            UserRole.GUEST: [
                Permission.READ
            ]
        }
    
    def check_permission(self, user: User, permission: Permission) -> bool:
        """Check if user has specific permission."""
        for role in user.roles:
            if permission in self.role_permissions.get(role, []):
                return True
        return False
    
    def get_user_permissions(self, user: User) -> List[Permission]:
        """Get all permissions for a user."""
        permissions = set()
        for role in user.roles:
            permissions.update(self.role_permissions.get(role, []))
        return list(permissions)
    
    def add_role_permission(self, role: UserRole, permission: Permission):
        """Add permission to a role."""
        if role not in self.role_permissions:
            self.role_permissions[role] = []
        if permission not in self.role_permissions[role]:
            self.role_permissions[role].append(permission)
    
    def remove_role_permission(self, role: UserRole, permission: Permission):
        """Remove permission from a role."""
        if role in self.role_permissions and permission in self.role_permissions[role]:
            self.role_permissions[role].remove(permission)


class ABACService:
    """Attribute-Based Access Control implementation."""
    
    def __init__(self):
        self.policies = []
    
    def add_policy(self, policy: Dict[str, Any]):
        """Add an ABAC policy."""
        self.policies.append(policy)
    
    def check_access(self, user: User, resource: Resource, action: str, context: Dict[str, Any] = None) -> bool:
        """Check access using ABAC policies."""
        context = context or {}
        
        for policy in self.policies:
            if self._evaluate_policy(policy, user, resource, action, context):
                return policy.get('effect', 'deny') == 'allow'
        
        return False  # Default deny
    
    def _evaluate_policy(self, policy: Dict[str, Any], user: User, resource: Resource, action: str, context: Dict[str, Any]) -> bool:
        """Evaluate a single ABAC policy."""
        conditions = policy.get('conditions', {})
        
        # Check user attributes
        for attr, value in conditions.get('user_attributes', {}).items():
            if user.attributes.get(attr) != value:
                return False
        
        # Check resource attributes
        for attr, value in conditions.get('resource_attributes', {}).items():
            if resource.attributes.get(attr) != value:
                return False
        
        # Check action
        if 'action' in conditions and action not in conditions['action']:
            return False
        
        # Check context
        for key, value in conditions.get('context', {}).items():
            if context.get(key) != value:
                return False
        
        return True
    
    def create_policy(self, name: str, effect: str, conditions: Dict[str, Any]) -> Dict[str, Any]:
        """Create a new ABAC policy."""
        return {
            'name': name,
            'effect': effect,
            'conditions': conditions
        }


class SessionManager:
    """Session management for security."""
    
    def __init__(self):
        self.active_sessions = {}
        self.session_timeout = timedelta(hours=8)
    
    def create_session(self, user_id: str, session_data: Dict[str, Any]) -> str:
        """Create a new session."""
        session_id = secrets.token_urlsafe(32)
        session = {
            'user_id': user_id,
            'data': session_data,
            'created_at': datetime.utcnow(),
            'last_activity': datetime.utcnow()
        }
        self.active_sessions[session_id] = session
        return session_id
    
    def get_session(self, session_id: str) -> Optional[Dict[str, Any]]:
        """Get session data."""
        session = self.active_sessions.get(session_id)
        if not session:
            return None
        
        # Check session timeout
        if datetime.utcnow() - session['last_activity'] > self.session_timeout:
            self.remove_session(session_id)
            return None
        
        # Update last activity
        session['last_activity'] = datetime.utcnow()
        return session
    
    def remove_session(self, session_id: str):
        """Remove a session."""
        if session_id in self.active_sessions:
            del self.active_sessions[session_id]
    
    def cleanup_expired_sessions(self):
        """Clean up expired sessions."""
        current_time = datetime.utcnow()
        expired_sessions = [
            session_id for session_id, session in self.active_sessions.items()
            if current_time - session['last_activity'] > self.session_timeout
        ]
        for session_id in expired_sessions:
            self.remove_session(session_id)


class MultiFactorAuth:
    """Multi-factor authentication implementation."""
    
    def __init__(self):
        self.totp_secrets = {}
    
    def generate_totp_secret(self, user_id: str) -> str:
        """Generate TOTP secret for user."""
        secret = secrets.token_urlsafe(32)
        self.totp_secrets[user_id] = secret
        return secret
    
    def verify_totp(self, user_id: str, totp_code: str) -> bool:
        """Verify TOTP code."""
        secret = self.totp_secrets.get(user_id)
        if not secret:
            return False
        
        # Simple TOTP verification (in production, use proper TOTP library)
        expected_code = self._generate_totp_code(secret)
        return totp_code == expected_code
    
    def _generate_totp_code(self, secret: str) -> str:
        """Generate TOTP code (simplified implementation)."""
        # In production, use proper TOTP implementation
        import time
        timestamp = int(time.time() // 30)
        hash_input = f"{secret}{timestamp}".encode()
        hash_result = hashlib.sha256(hash_input).hexdigest()
        return hash_result[:6]  # 6-digit code 

File: arx-backend/services/security/authentication.go
package security

import (
	"crypto/rand"
	"encoding/base64"
	"fmt"
	"sync"
	"time"

	"github.com/golang-jwt/jwt/v5"
	"github.com/google/uuid"
	"golang.org/x/crypto/bcrypt"
)

// UserRole represents user roles for RBAC
type UserRole string

const (
	RoleAdmin      UserRole = "admin"
	RoleEngineer   UserRole = "engineer"
	RoleViewer     UserRole = "viewer"
	RoleContractor UserRole = "contractor"
	RoleGuest      UserRole = "guest"
)

// Permission represents permissions for resource access control
type Permission string

const (
	PermissionRead          Permission = "read"
	PermissionWrite         Permission = "write"
	PermissionDelete        Permission = "delete"
	PermissionManageUsers   Permission = "manage_users"
	PermissionManageSystem  Permission = "manage_system"
	PermissionExportData    Permission = "export_data"
	PermissionImportData    Permission = "import_data"
	PermissionViewAuditLogs Permission = "view_audit_logs"
)

// User represents a user entity with attributes for ABAC
type User struct {
	ID         string                 `json:"id"`
	Username   string                 `json:"username"`
	Email      string                 `json:"email"`
	Roles      []UserRole             `json:"roles"`
	Attributes map[string]interface{} `json:"attributes"`
	IsActive   bool                   `json:"is_active"`
	CreatedAt  time.Time              `json:"created_at"`
	LastLogin  *time.Time             `json:"last_login,omitempty"`
}

// Resource represents a resource entity for access control
type Resource struct {
	ID         string                 `json:"id"`
	Type       string                 `json:"type"`
	Attributes map[string]interface{} `json:"attributes"`
	OwnerID    string                 `json:"owner_id"`
	CreatedAt  time.Time              `json:"created_at"`
}

// Session represents a user session
type Session struct {
	ID           string                 `json:"id"`
	UserID       string                 `json:"user_id"`
	Data         map[string]interface{} `json:"data"`
	CreatedAt    time.Time              `json:"created_at"`
	LastActivity time.Time              `json:"last_activity"`
	ExpiresAt    time.Time              `json:"expires_at"`
}

// TokenClaims represents JWT token claims
type TokenClaims struct {
	UserID   string     `json:"user_id"`
	Username string     `json:"username"`
	Roles    []UserRole `json:"roles"`
	jwt.RegisteredClaims
}

// AuthService provides authentication and authorization services
type AuthService struct {
	secretKey          string
	algorithm          string
	tokenExpiry        time.Duration
	refreshTokenExpiry time.Duration
	sessionTimeout     time.Duration
	activeSessions     map[string]*Session
	sessionMutex       sync.RWMutex
	rolePermissions    map[UserRole][]Permission
	rbacMutex          sync.RWMutex
}

// NewAuthService creates a new authentication service
func NewAuthService(secretKey string) *AuthService {
	as := &AuthService{
		secretKey:          secretKey,
		algorithm:          "HS256",
		tokenExpiry:        24 * time.Hour,
		refreshTokenExpiry: 7 * 24 * time.Hour,
		sessionTimeout:     8 * time.Hour,
		activeSessions:     make(map[string]*Session),
		rolePermissions:    make(map[UserRole][]Permission),
	}

	// Initialize default role permissions
	as.initializeRolePermissions()

	return as
}

// initializeRolePermissions sets up default role permissions
func (as *AuthService) initializeRolePermissions() {
	as.rbacMutex.Lock()
	defer as.rbacMutex.Unlock()

	as.rolePermissions[RoleAdmin] = []Permission{
		PermissionRead, PermissionWrite, PermissionDelete,
		PermissionManageUsers, PermissionManageSystem,
		PermissionExportData, PermissionImportData,
		PermissionViewAuditLogs,
	}

	as.rolePermissions[RoleEngineer] = []Permission{
		PermissionRead, PermissionWrite, PermissionExportData,
		PermissionImportData,
	}

	as.rolePermissions[RoleViewer] = []Permission{
		PermissionRead,
	}

	as.rolePermissions[RoleContractor] = []Permission{
		PermissionRead, PermissionWrite,
	}

	as.rolePermissions[RoleGuest] = []Permission{
		PermissionRead,
	}
}

// HashPassword hashes a password using bcrypt
func (as *AuthService) HashPassword(password string) (string, error) {
	hashedBytes, err := bcrypt.GenerateFromPassword([]byte(password), bcrypt.DefaultCost)
	if err != nil {
		return "", fmt.Errorf("failed to hash password: %w", err)
	}
	return string(hashedBytes), nil
}

// VerifyPassword verifies a password against its hash
func (as *AuthService) VerifyPassword(password, hashedPassword string) bool {
	err := bcrypt.CompareHashAndPassword([]byte(hashedPassword), []byte(password))
	return err == nil
}

// GenerateToken generates a JWT token for a user
func (as *AuthService) GenerateToken(user *User) (string, error) {
	now := time.Now()
	claims := &TokenClaims{
		UserID:   user.ID,
		Username: user.Username,
		Roles:    user.Roles,
		RegisteredClaims: jwt.RegisteredClaims{
			ExpiresAt: jwt.NewNumericDate(now.Add(as.tokenExpiry)),
			IssuedAt:  jwt.NewNumericDate(now),
			NotBefore: jwt.NewNumericDate(now),
			Issuer:    "arxos",
			Subject:   user.ID,
			ID:        uuid.New().String(),
		},
	}

	token := jwt.NewWithClaims(jwt.SigningMethodHS256, claims)
	return token.SignedString([]byte(as.secretKey))
}

// GenerateRefreshToken generates a refresh token
func (as *AuthService) GenerateRefreshToken(user *User) (string, error) {
	now := time.Now()
	claims := &TokenClaims{
		UserID:   user.ID,
		Username: user.Username,
		Roles:    user.Roles,
		RegisteredClaims: jwt.RegisteredClaims{
			ExpiresAt: jwt.NewNumericDate(now.Add(as.refreshTokenExpiry)),
			IssuedAt:  jwt.NewNumericDate(now),
			NotBefore: jwt.NewNumericDate(now),
			Issuer:    "arxos",
			Subject:   user.ID,
			ID:        uuid.New().String(),
		},
	}

	token := jwt.NewWithClaims(jwt.SigningMethodHS256, claims)
	return token.SignedString([]byte(as.secretKey))
}

// VerifyToken verifies and decodes a JWT token
func (as *AuthService) VerifyToken(tokenString string) (*TokenClaims, error) {
	token, err := jwt.ParseWithClaims(tokenString, &TokenClaims{}, func(token *jwt.Token) (interface{}, error) {
		if _, ok := token.Method.(*jwt.SigningMethodHMAC); !ok {
			return nil, fmt.Errorf("unexpected signing method: %v", token.Header["alg"])
		}
		return []byte(as.secretKey), nil
	})

	if err != nil {
		return nil, fmt.Errorf("failed to parse token: %w", err)
	}

	if claims, ok := token.Claims.(*TokenClaims); ok && token.Valid {
		return claims, nil
	}

	return nil, fmt.Errorf("invalid token")
}

// RefreshToken generates a new access token from a refresh token
func (as *AuthService) RefreshToken(refreshToken string) (string, error) {
	claims, err := as.VerifyToken(refreshToken)
	if err != nil {
		return "", fmt.Errorf("invalid refresh token: %w", err)
	}

	// Create user object from claims
	user := &User{
		ID:       claims.UserID,
		Username: claims.Username,
		Roles:    claims.Roles,
	}

	return as.GenerateToken(user)
}

// CreateSession creates a new session for a user
func (as *AuthService) CreateSession(userID string, sessionData map[string]interface{}) (*Session, error) {
	sessionID, err := generateSessionID()
	if err != nil {
		return nil, fmt.Errorf("failed to generate session ID: %w", err)
	}

	now := time.Now()
	session := &Session{
		ID:           sessionID,
		UserID:       userID,
		Data:         sessionData,
		CreatedAt:    now,
		LastActivity: now,
		ExpiresAt:    now.Add(as.sessionTimeout),
	}

	as.sessionMutex.Lock()
	as.activeSessions[sessionID] = session
	as.sessionMutex.Unlock()

	return session, nil
}

// GetSession retrieves session data
func (as *AuthService) GetSession(sessionID string) (*Session, error) {
	as.sessionMutex.RLock()
	session, exists := as.activeSessions[sessionID]
	as.sessionMutex.RUnlock()

	if !exists {
		return nil, fmt.Errorf("session not found")
	}

	// Check if session has expired
	if time.Now().After(session.ExpiresAt) {
		as.RemoveSession(sessionID)
		return nil, fmt.Errorf("session has expired")
	}

	// Update last activity
	as.sessionMutex.Lock()
	session.LastActivity = time.Now()
	session.ExpiresAt = time.Now().Add(as.sessionTimeout)
	as.sessionMutex.Unlock()

	return session, nil
}

// RemoveSession removes a session
func (as *AuthService) RemoveSession(sessionID string) {
	as.sessionMutex.Lock()
	delete(as.activeSessions, sessionID)
	as.sessionMutex.Unlock()
}

// CleanupExpiredSessions removes expired sessions
func (as *AuthService) CleanupExpiredSessions() {
	as.sessionMutex.Lock()
	defer as.sessionMutex.Unlock()

	now := time.Now()
	for sessionID, session := range as.activeSessions {
		if now.After(session.ExpiresAt) {
			delete(as.activeSessions, sessionID)
		}
	}
}

// CheckPermission checks if a user has a specific permission
func (as *AuthService) CheckPermission(user *User, permission Permission) bool {
	as.rbacMutex.RLock()
	defer as.rbacMutex.RUnlock()

	for _, role := range user.Roles {
		if permissions, exists := as.rolePermissions[role]; exists {
			for _, perm := range permissions {
				if perm == permission {
					return true
				}
			}
		}
	}
	return false
}

// GetUserPermissions returns all permissions for a user
func (as *AuthService) GetUserPermissions(user *User) []Permission {
	as.rbacMutex.RLock()
	defer as.rbacMutex.RUnlock()

	permissions := make(map[Permission]bool)
	for _, role := range user.Roles {
		if rolePerms, exists := as.rolePermissions[role]; exists {
			for _, perm := range rolePerms {
				permissions[perm] = true
			}
		}
	}

	result := make([]Permission, 0, len(permissions))
	for perm := range permissions {
		result = append(result, perm)
	}
	return result
}

// AddRolePermission adds a permission to a role
func (as *AuthService) AddRolePermission(role UserRole, permission Permission) {
	as.rbacMutex.Lock()
	defer as.rbacMutex.Unlock()

	if as.rolePermissions[role] == nil {
		as.rolePermissions[role] = make([]Permission, 0)
	}

	// Check if permission already exists
	for _, existingPerm := range as.rolePermissions[role] {
		if existingPerm == permission {
			return
		}
	}

	as.rolePermissions[role] = append(as.rolePermissions[role], permission)
}

// RemoveRolePermission removes a permission from a role
func (as *AuthService) RemoveRolePermission(role UserRole, permission Permission) {
	as.rbacMutex.Lock()
	defer as.rbacMutex.Unlock()

	if permissions, exists := as.rolePermissions[role]; exists {
		for i, perm := range permissions {
			if perm == permission {
				as.rolePermissions[role] = append(permissions[:i], permissions[i+1:]...)
				break
			}
		}
	}
}

// ValidateUserCredentials validates user credentials
func (as *AuthService) ValidateUserCredentials(username, password string, hashedPassword string) bool {
	// Verify password
	if !as.VerifyPassword(password, hashedPassword) {
		return false
	}

	// Additional validation can be added here (e.g., check if user is active)
	return true
}

// generateSessionID generates a secure session ID
func generateSessionID() (string, error) {
	bytes := make([]byte, 32)
	if _, err := rand.Read(bytes); err != nil {
		return "", err
	}
	return base64.URLEncoding.EncodeToString(bytes), nil
}

// GetActiveSessionsCount returns the number of active sessions
func (as *AuthService) GetActiveSessionsCount() int {
	as.sessionMutex.RLock()
	defer as.sessionMutex.RUnlock()
	return len(as.activeSessions)
}

// GetSessionStats returns session statistics
func (as *AuthService) GetSessionStats() map[string]interface{} {
	as.sessionMutex.RLock()
	defer as.sessionMutex.RUnlock()

	now := time.Now()
	activeCount := 0
	expiredCount := 0

	for _, session := range as.activeSessions {
		if now.Before(session.ExpiresAt) {
			activeCount++
		} else {
			expiredCount++
		}
	}

	return map[string]interface{}{
		"total_sessions":   len(as.activeSessions),
		"active_sessions":  activeCount,
		"expired_sessions": expiredCount,
		"session_timeout":  as.sessionTimeout.String(),
	}
}
